site-packages/mlab/disco/collectd.py

```diff
@@ -0,0 +1,134 @@
+              pair was already added.
+        """
+        collection_id, data_block = self._format_data_block(
+            port_name, oid_group_name, oids)
+        if collection_id in self._collections:
+            raise DuplicateName('Collection name already added: %s' %
+                                collection_id)
+        self._collections.add(collection_id)
+        self._data_blocks.append(data_block)
+
+    def generate(self):
+        """Returns a collectd-snmp plugin configuration.
+
+        The plugin configuration only needs to be generated once, but may be
+        called multiple times.
+
+        Returns:
+          str, a collectd-snmp plugin configuration.
+        """
+        return self._format_snmp_plugin(self._format_host_block())
+
+    def _format_host_block(self):
+        """Creates a collectd-snmp configuration host block.
+
+        Returns:
+          str, a collectd-nsmp <Host></Host> block.
+        """
+        # Host name is used to name the output files.
+        data = '<Host "%s">\n' % self._hostname
+        # Address is used to contact the switch SNMP server.
+        data += '  Address "%s"\n' % self._hostname
+        data += '  Version 2\n'
+        data += '  Community "%s"\n' % self._community
+        data += '  Collect %s\n' % ' '.join(self._collections)
+        data += '  Interval %s\n' % self._interval
+        data += '</Host>'
+        return data
+
+    def _format_data_block(self, port_name, oid_group_name, oids):
+        """Formats a collectd-snmp configuration data block.
+
+        All data blocks should have a unique combination of (port_name,
+        oid_group_name).
+
+        Args:
+          port_name: str, name of switch port, e.g. 'uplink'.
+          oid_group_name: str, name of the oid to format, e.g. 'discards'.
+          oids: iterable of str, the actual numeric or symbolic OIDs to collect.
+
+        Returns:
+          tuple of (str, str), the first element is a unique collection name,
+              and the second is a collectd-snmp <Data></Data> block.
+        """
+        unique_name = _get_unique_name(port_name, oid_group_name)
+
+        data = '<Data "%s">\n' % unique_name
+        data += '  Type "%s"\n' % _get_collectd_type(oid_group_name)
+        data += '  Table false\n'
+        data += '  Instance "%s"\n' % port_name
+        data += '  Values ' + ' '.join(['"%s"' % v for v in oids]) + '\n'
+        data += '</Data>'
+        return unique_name, data
+
+    def _format_snmp_plugin(self, host_block):
+        """Formats a collectd-snmp plugin configuration.
+
+        Args:
+          host_block: str, a host block as created by format_snmp_host.
+
+        Returns:
+          str, the collectd-snmp configuration.
+        """
+        data = 'LoadPlugin snmp\n'
+        data += '<Plugin snmp>\n'
+        data += '\n'.join(self._data_blocks) + '\n'
+        data += '\n' + host_block + '\n'
```

Comment:
Okay. I added extra spaces to the "data" and "host" blocks to get the nested indent. See the sample in collectd_test.py for the net effect.